source/slang/slang-check-decl.cpp

```diff
@@ -1761,15 +1760,200 @@ namespace Slang
         }
     }
 
-    void SemanticsDeclBasesVisitor::visitAggTypeDecl(AggTypeDecl* decl)
+    void SemanticsDeclBasesVisitor::_validateCrossModuleInheritance(
+        AggTypeDeclBase* decl,
+        InheritanceDecl* inheritanceDecl)
     {
-        // TODO: We need to enumerate the bases here,
-        // and ideally form a "class precedence list"
-        // from them.
+        // Within a single module, users should be allowed to inherit
+        // one type from another more or less freely, so long as they
+        // don't violate fundamental validity conditions around
+        // inheritance.
+        //
+        // When an inheritance relationship is declared in one module,
+        // and the base type is in another module, we may want to
+        // enforce more restrictions. As a strong example, we probably
+        // don't want people to declare their own subtype of `int`
+        // or `Texture2D<float4>`.
+        //
+        // We start by checking if the type being inherited from is
+        // a decl-ref type, since that means it refers to a declaration
+        // that can be localized to its original module.
+        //
+        auto baseType = inheritanceDecl->base.type;
+        auto baseDeclRefType = as<DeclRefType>(baseType);
+        if( !baseDeclRefType )
+        {
+            return;
+        }
+        auto baseDecl = baseDeclRefType->declRef.decl;
+
+        // Using the parent/child hierarchy baked into `Decl`s we
+        // can find the modules that contain both the `decl` doing
+        // the inheriting, and the `baseDeclRefType` that is being
+        // inherited from.
+        //
+        // If those modules are the same, then we aren't seeing any
+        // kind of cross-module inheritance here, and there is nothing
+        // that needs enforcing.
+        //
+        auto moduleWithInheritance = getModule(decl);
+        auto moduleWithBaseType = getModule(baseDecl);
+        if( moduleWithInheritance == moduleWithBaseType )
+        {
+            return;
+        }
+
+        if( baseDecl->hasModifier<SealedAttribute>() )
+        {
+            // If the original declaration had the `[sealed]` attribute on it,
+            // then it explicitly does *not* allow inheritance from other
+            // modules.
+            //
+            getSink()->diagnose(inheritanceDecl, Diagnostics::cannotInheritFromExplicitlySealedDeclarationInAnotherModule, baseType, moduleWithBaseType->getModuleDecl()->getName());
```

Comment:
> Do we want to allow inheriting from a sealed type within a same module?

I would claim that we do, but I suppose that could be confusing to users at first.

There are really three cases worth supporting:

1. The type can be inherited from without restrictions (currently `[open]`)
2. The type can be inherited from, but only within the current module (currently `[sealed]`)
3. The type can't be inherited from at all (no syntax right now)

I would like to argue against (3) for the same reasons I'd be inclined to argue against a lexically-scoped `private` keyword: if the code is in *your* module, then you should be trusted to have whatever access to it you need. I realize that sentiment isn't universal.

Note that case (2) *must* be supported, because we need it in the Slang stdlib already: we have the interfaces like `__BuiltinFloatingPointType` such that a type like `float` needs to inherit from it, but it would lead to Very Bad Things if we let user types inherit.

If we had to support (3) as well then I'd claim the obvious name to use would be something akin to `[final]`, although one concern I have is that I'd rather not conflate the "ability to inherit" axis with the "ability to overide" axis if we start adding more OOP features (because conflating them makes things like virtual types/classes much harder to design).

I'd be happy to bikeshed the exact set of attributes and their spellings, and this change is just trying to establish a starting point that we can build from.

Reply:
I agree that we need case 2. This use of [sealed] to mean this could be a little confusing, but I think it is fine too.

Reply:
There is some precedent in other languages for `sealed` referring to the sealing of an abstraction at the module boundary (while something like `final` is explicit about the given declaration being the last word).

Can we come up with a better term that suggests the semantics more clearly, or is this something we should leave to documentation?